issue-26744-CLI-Implement-a-global-pull-for-files

jgambarios opened the pull request:
Improving the pull command for files, now it is possible to pull all files for all sites if "path" is not specified, also, the file’s pull was integrated into the global pull command.

jgambarios committed:
#26744 Supporting files pull into the global pull

jgambarios committed:
Merge remote-tracking branch 'origin/master' into issue-26744-CLI-Implement-a-global-pull-for-files

jgambarios committed:
#26744 Pull command improvements

jgambarios committed:
#26744 Refactor pull operations for separation of concerns and extensibility

Lots of changes have taken place to enhance the modularity and extensibility of the pull methods. Primarily, the pull operations, previously found in FilePullHandler and FilePullController, have now been encapsulated into a separate Puller class for better separation of concerns. This introduces an extra level of abstraction, making the code easier to manage and extend.

The pullTree and pullFile methods from the PullService have also been replaced by a single, more flexible pull method. Additionally, the processing of the file tree, which was initially done in the processTree method of the FilePullHandler, has been moved to the Puller class. This further enhances modularity and improves the overall code structure.

Test files have also been amended to accommodate for these changes. The refactor enhances the code maintainability and allows for easier enhancements and bug fixes in the future.

jgambarios committed:
#26744 Improving user feedback

jgambarios committed:
#26744 Improving user feedback on console

jgambarios committed:
#26744 Creating IT to test the pulling of all the existing sites.

jgambarios committed:
#26744 Small cleanup

jgambarios committed:
#26744 Small cleanup

jgambarios committed:
#26744 Improvements

jgambarios committed:
Merge remote-tracking branch 'origin/master' into issue-26744-CLI-Implement-a-global-pull-for-files

fabrizzio-dotCMS commented on tools/dotcms-cli/cli/src/test/java/com/dotcms/api/client/files/PushServiceIT.java:
```diff
@@ -438,21 +490,34 @@ void Test_Delete_Folder() throws IOException {
 
-            pullService.pullTree(outputOptions, result.getRight(), absolutePath.toFile(),
-                    true, true, true, 0);
+            Map<String, Object> customOptions = Map.of(
```
shouldn't these custom options be part of the builder?
I think the other should be part of the builder too

jgambarios replied:
Good idea, and done, now `customOptions` is a property of the `PullOptions`.

fabrizzio-dotCMS commented on tools/dotcms-cli/cli/src/test/java/com/dotcms/api/client/files/PushServiceIT.java:
```diff
@@ -280,21 +319,34 @@ void Test_Push_Modified_Data() throws IOException {
-            pullService.pullTree(outputOptions, result.getRight(), absolutePath.toFile(),
-                    true, true, true, 0);
+
+            Map<String, Object> customOptions = Map.of(
```
shouldn't these custom options be part of the builder?

jgambarios replied:
Good idea, and done, now `customOptions` is a property of the `PullOptions`.

fabrizzio-dotCMS commented on tools/dotcms-cli/cli/src/test/java/com/dotcms/api/client/files/PushServiceIT.java:
```diff
@@ -96,20 +109,33 @@ void Test_Nothing_To_Push() throws IOException {
+            );
+
+            // Execute the pull
+            pullService.pull(
```
I would change this to only take the PullOptions any additional param could make it into the PullOptions

jgambarios replied:
If that's OK, I will leave the `PullOptions` as the pull commands options holder but not as holder for the method parameters, but following your advice I added inside the `PullOptions` the `customOptions`.

jgambarios committed:
Merge remote-tracking branch 'origin/master' into issue-26744-CLI-Implement-a-global-pull-for-files

# Conflicts:
#	tools/dotcms-cli/cli/src/test/java/com/dotcms/api/client/files/PullServiceIT.java
#	tools/dotcms-cli/cli/src/test/java/com/dotcms/api/client/files/PushServiceIT.java

jgambarios committed:
#26744 Applying feedback

jgambarios commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/GenericPullHandler.java:
```diff
@@ -0,0 +1,24 @@
+ *
+ * @param <T>
+ */
+public interface GenericPullHandler<T> extends PullHandler<T> {
```
Sounds better thanks, done.

jdotcms commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/CustomPullHandler.java:
```diff
@@ -0,0 +1,33 @@
+     * @param output        The output options for pulling the content.
+     * @return A list of exceptions encountered during the pull operation.
+     */
+    List<Exception> pull(
```
why returning an exceptions, it looks odd

jdotcms commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/PullHandler.java:
```diff
@@ -14,20 +18,13 @@ public interface PullHandler<T> {
      * Returns a short format of a given T element. Used for console user feedback.
      */
-    String shortFormat(T content);
+    String shortFormat(T content, PullOptions pullOptions, Map<String, Object> customOptions);
```
might be customOptions could be a property into the pull options to avoid a lot of arguments

jgambarios replied:
Good idea, and done, now `customOptions` is a property of the `PullOptions`.

jdotcms commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/PullServiceImpl.java:
```diff
@@ -169,39 +166,205 @@ private <T> List<T> fetch(
+
+        output.info(pullHandler.startPullingHeader(contents));
+
+        for (var content : contents) {
```
set to final

jgambarios replied:
Done

jdotcms commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/PullServiceImpl.java:
```diff
@@ -210,47 +373,81 @@ private <T> void pullToDisk(File destination,
         );
 
-        CompletableFuture<Void> pullFuture = CompletableFuture.supplyAsync(
+        CompletableFuture<List<Exception>> pullFuture = CompletableFuture.supplyAsync(
```
wondering why you need thread in a single process app

jdotcms commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/PullServiceImpl.java:
```diff
@@ -210,47 +373,81 @@ private <T> void pullToDisk(File destination,
+                    )
+            );
+            for (var error : errors) {
+                if (error instanceof TraversalTaskException || error instanceof PullException) {
```
this seems to be a case for strategy + polymorphism

jdotcms commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/contenttype/ContentTypePullHandler.java:
```diff
@@ -1,19 +1,31 @@
     }
 
+    @Override
+    public String startPullingHeader(List<ContentType> contents) {
```
set to final

jgambarios replied:
Done

jdotcms commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/file/FileFetcher.java:
```diff
@@ -0,0 +1,184 @@
+
+        // ---
+        // And now for each site we need to fetch the tree under the specified site
+        List<FileTraverseResult> results = new ArrayList<>();
```
set to final

jgambarios replied:
Done

jdotcms commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/file/FileFetcher.java:
```diff
@@ -0,0 +1,184 @@
+     * @return The set of values associated with the key, or null if the options map is null or does
+     * not contain the key
+     */
+    private Set<String> getSetFromOptions(Map<String, Object> options, String key) {
```
set to final

jdotcms replied:
this could be reduced to something such as

return Optiona.ofNullable(options). ifPresentOrElse(map -> options.getOrDefault(key, Collections.emptySet(), ()->Collections.emptySet());

jdotcms commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/file/FilePullHandler.java:
```diff
@@ -0,0 +1,262 @@
+import org.jboss.logging.Logger;
+
+@Dependent
+public class FilePullHandler implements CustomPullHandler<FileTraverseResult> {
```
Doc

jgambarios replied:
Done

jdotcms commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/file/FilePullHandler.java:
```diff
@@ -0,0 +1,262 @@
+        final TreeNode tree = treeInfo.getLeft();
+        final TreeNodeInfo treeNodeInfo = treeInfo.getRight();
+
+        StringBuilder sb = new StringBuilder();
```
final

jdotcms commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/file/FilePullHandler.java:
```diff
@@ -0,0 +1,262 @@
+
+    @Override
+    @ActivateRequestContext
+    public List<Exception> pull(FileTraverseResult content, final PullOptions pullOptions,
```
final

jgambarios replied:
Done

jdotcms commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/file/OptionConstants.java:
```diff
@@ -0,0 +1,20 @@
+package com.dotcms.api.client.pull.file;
+
+public class OptionConstants {
```
Doc

jgambarios replied:
Done

jdotcms commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/task/AbstractPullTaskParams.java:
```diff
@@ -35,4 +35,6 @@ public interface AbstractPullTaskParams<T> extends Serializable {
 
     ConsoleProgressBar progressBar();
 
+    boolean failFast();
```
doc

jgambarios replied:
Done

jgambarios committed:
#26744 Applying feedback

jgambarios committed:
#26744 Applying feedback

jgambarios commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/file/FileFetcher.java:
```diff
@@ -0,0 +1,184 @@
+        // ---
+        // And now for each site we need to fetch the tree under the specified site
+        List<FileTraverseResult> results = new ArrayList<>();
+        for (Site site : allSites) {
```
Done

jgambarios commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/file/FileFetcher.java:
```diff
@@ -0,0 +1,184 @@
+
+    @ActivateRequestContext
+    @Override
+    public List<FileTraverseResult> fetch(Map<String, Object> customOptions) {
```
Done

jgambarios commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/file/AbstractFileTraverseResult.java:
```diff
@@ -0,0 +1,23 @@
+@Value.Immutable
+public interface AbstractFileTraverseResult {
+
+    List<Exception> exceptions();
```
Done

jgambarios committed:
#26744 Applying feedback

jgambarios commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/PullServiceImpl.java:
```diff
@@ -210,47 +373,81 @@ private <T> void pullToDisk(File destination,
+                            errors.size()
+                    )
+            );
+            for (var error : errors) {
```
Done

jgambarios commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/PullServiceImpl.java:
```diff
@@ -210,47 +373,81 @@ private <T> void pullToDisk(File destination,
+     * @param errors The list of errors to be printed.
+     * @param output The output option mixin for providing output messages.
+     */
+    private void printErrors(List<Exception> errors, OutputOptionMixin output) {
```
Done

jgambarios committed:
Merge remote-tracking branch 'origin/master' into issue-26744-CLI-Implement-a-global-pull-for-files

jgambarios committed:
#26744 "Refactor fetching sites with SiteIterator"

Refactored site data fetching process by introducing SiteIterator. Important parts like site fetching in FileFetcher and SiteFetcher methods get significantly simplified, increasing maintainability and readability. SiteIterator separately handles paging logic, which offers a cleaner and more concise site retrieval process.

jgambarios committed:
#26744 Moved SiteIterator from the 'common' package to a 'site' package.

jgambarios committed:
#26744 "Refactor PullHandler implementations to represent clearly defined strategies

This commit refactors PullHandler and its subclasses, namely ContentTypePullHandler, LanguagePullHandler, and SitePullHandler. The original structure was redefined, shifting from an interface to an abstract class.
The transformative intent is to reshape PullHandler into a true strategy pattern implementation. Subsequently, CustomPullHandler was deleted and its responsibilities were assumed by the abstract PullHandler class. Following this, all implementations of PullHandler have been modified to specifically extend this class, ensuring more coherent structure and clarity relating to their individual functionalities."

jgambarios committed:
#26744 Add error handling strategies for exceptions in PullHandler

Introduced ErrorPrinterStrategy interface to handle various types of exceptions with distinctive strategies. Added strategies for TraversalTaskException, PullException, and default exceptions. This new approach makes error handling cleaner and more extensible by decoupling the error handling logic from PullHandler, which is responsible for pulling elements. This is particularly useful for adding new types of exceptions in the future or changing the formatting of error messages without modifying the core PullHandler code.

jgambarios committed:
Merge remote-tracking branch 'origin/master' into issue-26744-CLI-Implement-a-global-pull-for-files

jgambarios commented on tools/dotcms-cli/cli/src/main/java/com/dotcms/api/client/pull/PullServiceImpl.java:
```diff
@@ -169,39 +166,205 @@ private <T> List<T> fetch(
+
+            boolean foundErrors;
+
+            if (pullHandler instanceof CustomPullHandler) {
```
As each implementation of the `PullHandler` can choose only between using a default implementation of the "pull" or to provide a custom "pull" implementation, I refactored and improved the code to handle this case using abstract classes, I think now is more clear how it works.
